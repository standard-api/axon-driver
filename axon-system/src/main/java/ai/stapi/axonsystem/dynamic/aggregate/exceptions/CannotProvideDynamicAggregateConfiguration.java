package ai.stapi.axonsystem.dynamic.aggregate.exceptions;

public class CannotProvideDynamicAggregateConfiguration extends Exception {

  public CannotProvideDynamicAggregateConfiguration(String aggregateType) {
    super(
        String.format(
            "Cannot provide Dynamic Aggregate Configuration for '%s', because it does not exists.",
            aggregateType
        )
    );
  }

  public CannotProvideDynamicAggregateConfiguration(String aggregateType, Throwable cause) {
    super(
        String.format(
            "Cannot provide Dynamic Aggregate Configuration for '%s', because it does not exists.\n Cause: %s",
            aggregateType,
            cause.getMessage()
        ),
        cause
    );
  }
}
