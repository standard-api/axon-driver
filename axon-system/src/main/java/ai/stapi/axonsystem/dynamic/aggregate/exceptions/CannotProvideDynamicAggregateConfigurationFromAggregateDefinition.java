package ai.stapi.axonsystem.dynamic.aggregate.exceptions;

import ai.stapi.graphsystem.aggregatedefinition.model.CommandHandlerDefinitionDTO;
import ai.stapi.graphsystem.operationdefinition.model.OperationDefinitionDTO;

public class CannotProvideDynamicAggregateConfigurationFromAggregateDefinition
    extends RuntimeException {

  private CannotProvideDynamicAggregateConfigurationFromAggregateDefinition(String becauseMessage) {
    super("Cannot provide DynamicAggregateConfiguration from OperationDefinition, because "
        + becauseMessage);
  }

  public static CannotProvideDynamicAggregateConfigurationFromAggregateDefinition becauseOperationHasNoneOrMoreResources(
      OperationDefinitionDTO operationDefinitionDTO
  ) {
    return new CannotProvideDynamicAggregateConfigurationFromAggregateDefinition(
        "operation definition has none or more resources." +
            "\nOperation definition: " + operationDefinitionDTO
    );
  }

  public static CannotProvideDynamicAggregateConfigurationFromAggregateDefinition becauseCommandHandlerDefinitionHasUnknownCreationalPolicy(
      CommandHandlerDefinitionDTO commandDefinition
  ) {
    return new CannotProvideDynamicAggregateConfigurationFromAggregateDefinition(
        "command handler definition has unknown creational policy." +
            "\nCommand handler definition: " + commandDefinition
    );
  }
}
