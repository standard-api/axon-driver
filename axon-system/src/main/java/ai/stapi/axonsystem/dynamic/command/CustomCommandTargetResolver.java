package ai.stapi.axonsystem.dynamic.command;

import ai.stapi.graphsystem.messaging.command.AbstractCommand;
import org.axonframework.commandhandling.CommandMessage;
import org.axonframework.modelling.command.AnnotationCommandTargetResolver;
import org.axonframework.modelling.command.CommandTargetResolver;
import org.axonframework.modelling.command.VersionedAggregateIdentifier;
import org.jetbrains.annotations.NotNull;

public class CustomCommandTargetResolver implements CommandTargetResolver {
  
  private final AnnotationCommandTargetResolver annotationCommandTargetResolver;

  public CustomCommandTargetResolver() {
    this.annotationCommandTargetResolver = AnnotationCommandTargetResolver.builder().build();
  }

  @Override
  public VersionedAggregateIdentifier resolveTarget(@NotNull CommandMessage<?> command) {
    var payload = command.getPayload();
    if (payload instanceof AbstractCommand<?> abstractCommand) {
      return new VersionedAggregateIdentifier(
          abstractCommand.getTargetIdentifier(),
          null
      );
    }
    return this.annotationCommandTargetResolver.resolveTarget(command);
  }
}
