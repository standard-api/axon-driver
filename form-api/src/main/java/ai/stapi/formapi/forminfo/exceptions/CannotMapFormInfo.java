package ai.stapi.formapi.forminfo.exceptions;

public class CannotMapFormInfo extends RuntimeException {

    private static final String MSG = "Cannot map form info, because ";

    private CannotMapFormInfo(String becauseMessage) {
        super(MSG + becauseMessage);
    }

    private CannotMapFormInfo(String becauseMessage, Throwable cause) {
        super(MSG + becauseMessage, cause);
    }

    public static CannotMapFormInfo becauseThereWasNoCommandHandlerDefinitionForProvidedOperation(
        String operationId
    ) {
        return new CannotMapFormInfo(
            String.format(
                "there was not command handler definition with provided operation.%nOperation name: '%s'",
                operationId
            )
        );
    }

    public static CannotMapFormInfo becauseThereWasNoCommandHandlerDefinitionForProvidedOperation(
        String operationId,
        Throwable cause
    ) {
        return new CannotMapFormInfo(
            String.format(
                "there was not command handler definition with provided operation.%nOperation name: '%s'",
                operationId
            ),
            cause
        );
    }
}
