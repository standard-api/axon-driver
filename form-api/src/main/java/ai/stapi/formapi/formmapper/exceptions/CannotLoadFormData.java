package ai.stapi.formapi.formmapper.exceptions;

public class CannotLoadFormData extends RuntimeException {

    private static final String MSG = "Cannot load data for form, because ";

    private CannotLoadFormData(String becauseMessage) {
        super(MSG + becauseMessage);
    }

    private CannotLoadFormData(String becauseMessage, Throwable cause) {
        super(MSG + becauseMessage, cause);
    }

    public static CannotLoadFormData becauseThereWasNoCommandHandlerDefinitionForProvidedOperation(
        String operationId
    ) {
        return new CannotLoadFormData(
            String.format(
                "there was not command handler definition with provided operation.%nOperation name: '%s'",
                operationId
            )
        );
    }

    public static CannotLoadFormData becauseThereWasNoCommandHandlerDefinitionForProvidedOperation(
        String operationId,
        Throwable cause
    ) {
        return new CannotLoadFormData(
            String.format(
                "there was not command handler definition with provided operation.%nOperation name: '%s'",
                operationId
            ),
            cause
        );
    }

    public static CannotLoadFormData becauseAggregateByProvidedResourceIdWasNotFound(
        String resourceId,
        Throwable cause
    ) {
        return new CannotLoadFormData(
            String.format(
                "no aggregate with provided resource id was found.%nResource id: '%s'",
                resourceId
            ),
            cause
        );
    }
}
