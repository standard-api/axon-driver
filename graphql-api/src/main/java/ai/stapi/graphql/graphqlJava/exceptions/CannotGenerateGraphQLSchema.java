package ai.stapi.graphql.graphqlJava.exceptions;

public class CannotGenerateGraphQLSchema extends RuntimeException {

  private CannotGenerateGraphQLSchema(String becauseMessage) {
    super("Cannot generate graphQl schema, because " + becauseMessage);
  }

  public static CannotGenerateGraphQLSchema becauseStructureSchemaFieldHadNoType(
      String typeName,
      String fieldName
  ) {
    return new CannotGenerateGraphQLSchema(
        "structure schema has some field which has no type." +
            "\nType name: " + typeName +
            "\nField name: " + fieldName
    );
  }

  public static CannotGenerateGraphQLSchema becauseOperationReferencedTypeWhichDoesNotExistInStructureSchema(
      String operationName,
      String parameterTypeName
  ) {
    return new CannotGenerateGraphQLSchema(
        "operation has parameter of type which does not exist in structure schema." +
            "\nOperation name: " + operationName +
            "\nParameter Type name: " + parameterTypeName
    );
  }
}
