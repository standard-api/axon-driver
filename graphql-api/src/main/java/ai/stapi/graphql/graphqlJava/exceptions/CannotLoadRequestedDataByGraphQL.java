package ai.stapi.graphql.graphqlJava.exceptions;

import ai.stapi.schema.structureSchemaProvider.exception.CannotProvideStructureSchema;
import graphql.language.ObjectValue;
import graphql.language.Value;
import graphql.schema.GraphQLFieldDefinition;
import java.util.List;

public class CannotLoadRequestedDataByGraphQL extends RuntimeException {

  private CannotLoadRequestedDataByGraphQL(String becauseMessage) {
    super("Cannot load requested data by GraphQL, because " + becauseMessage);
  }

  private CannotLoadRequestedDataByGraphQL(String becauseMessage, Throwable e) {
    super("Cannot load requested data by GraphQL, because " + becauseMessage, e);
  }

  public static CannotLoadRequestedDataByGraphQL becauseInvalidFieldInQuery(
      GraphQLFieldDefinition fieldDefinition) {
    return new CannotLoadRequestedDataByGraphQL(
        "invalid field encountered in main Query object." +
            "\nField should either return list and take search query parameters. Or return object and take string id parameter."
            +
            "\nGQL field definition: " + fieldDefinition.toString()
    );
  }

  public static CannotLoadRequestedDataByGraphQL becauseInnerFieldWasNotObjectOrScalar(
      GraphQLFieldDefinition fieldDefinition) {
    return new CannotLoadRequestedDataByGraphQL(
        "some inner field definition was not Object or Scalar." +
            "\nDefinition: " + fieldDefinition.toString()
    );
  }

  public static CannotLoadRequestedDataByGraphQL becauseProvidedSortOptionDidNotHaveExactlyOneField(
      ObjectValue sortOptionValue) {
    return new CannotLoadRequestedDataByGraphQL(
        "provided sort option did not have exactly one field." +
            "\nSort option: " + sortOptionValue.toString()
    );
  }

  public static CannotLoadRequestedDataByGraphQL becauseProvidedFilterOptionDidNotHaveExactlyOneField(
      ObjectValue filterOptionValue) {
    return new CannotLoadRequestedDataByGraphQL(
        "provided filter option did not have exactly one field." +
            "\nFilter option: " + filterOptionValue.toString()
    );
  }

  public static CannotLoadRequestedDataByGraphQL becauseProvidedSortOptionWasOfUnknownType(
      Value<?> value) {
    return new CannotLoadRequestedDataByGraphQL(
        "provided sort option was of unknown type." +
            "\nSort value: " + value.toString()
    );
  }

  public static CannotLoadRequestedDataByGraphQL becauseProvidedFilterOptionWasOfUnknownStrategy(
      String filterStrategy) {
    return new CannotLoadRequestedDataByGraphQL(
        "provided filter option was of unknown strategy." +
            "\nStrategy: " + filterStrategy
    );
  }

  public static CannotLoadRequestedDataByGraphQL becauseProvidedListFilterInputWasNotOfObjectValue(
      String filterStrategy,
      String attributeName,
      Value<?> value
  ) {
    return new CannotLoadRequestedDataByGraphQL(
        "provided list filter input was not of object value." +
            "\nActual value: " + value +
            "\nStrategy: " + filterStrategy +
            "\nAttribute name: " + attributeName
    );
  }

  public static CannotLoadRequestedDataByGraphQL becauseProvidedMultiLogicalFilterInputWasNotOfArrayValue(
      String filterStrategy,
      Value<?> value
  ) {
    return new CannotLoadRequestedDataByGraphQL(
        "provided multi logical filter input was not of array value." +
            "\nActual value: " + value +
            "\nStrategy: " + filterStrategy
    );
  }

  public static CannotLoadRequestedDataByGraphQL becauseProvidedLogicalFilterInputHadInvalidType(
      String filterStrategy,
      List<Value> invalidValueItems
  ) {
    return new CannotLoadRequestedDataByGraphQL(
        "provided logical filter input was not object value." +
            "\nInvalid values: " + invalidValueItems +
            "\nStrategy: " + filterStrategy
    );
  }

  public static CannotLoadRequestedDataByGraphQL becauseProvidedDeepFilterWasNotOfObjectValue(
      String edgeType,
      Value<?> value
  ) {
    return new CannotLoadRequestedDataByGraphQL(
        "provided deep filter input was not of object value." +
            "\nActual value: " + value +
            "\nEdge type/Field name: " + edgeType
    );
  }

  public static CannotLoadRequestedDataByGraphQL becauseProvidedMutationDidNotHaveCorrespondingOperationDefinition(
      String commandName
  ) {
    return new CannotLoadRequestedDataByGraphQL(
        "provided mutation did not have corresponding operation definition." +
            "\nCommand name: " + commandName
    );
  }

  public static CannotLoadRequestedDataByGraphQL becauseProvidedMutationHadParameterOfUnknownType(
      String commandName,
      String parameterName,
      String parameterType
  ) {
    return new CannotLoadRequestedDataByGraphQL(
        "provided mutation had parameter of unknown type." +
            "\nCommand name: " + commandName +
            "\nParameter name: " + parameterName +
            "\nParameter type: " + parameterType
    );
  }

  public static CannotLoadRequestedDataByGraphQL becauseProvidedMutationHadParameterOfComplexTypeWithSomeFieldOfUnknownType(
      String commandName,
      String complexTypeName,
      String fieldName,
      String fieldType
  ) {
    return new CannotLoadRequestedDataByGraphQL(
        "provided mutation had parameter of complex type with some field of unknown type." +
            "\nCommand name: " + commandName +
            "\nType name: " + complexTypeName +
            "\nField name: " + fieldName +
            "\nField type: " + fieldType
    );
  }

  public static CannotLoadRequestedDataByGraphQL becauseProvidedMutationHadParameterOfComplexTypeWithSomeInvalidUnionValue(
      String commandName,
      String typeName,
      String fieldName,
      Object actualUnionValue
  ) {
    return new CannotLoadRequestedDataByGraphQL(
        "provided mutation had parameter of complex type with some field of invalid union type." +
            "\nCommand name: " + commandName +
            "\nType name: " + typeName +
            "\nField name: " + fieldName +
            "\nUnion value: " + actualUnionValue
    );
  }

  public static CannotLoadRequestedDataByGraphQL becauseUnionInputDidNotHaveExactlyOneMemberSpecified(
      String commandName,
      String typeName,
      String fieldName,
      Object actualUnionValue
  ) {
    return new CannotLoadRequestedDataByGraphQL(
        "provided mutation had parameter of complex type with some invalid field of union type." +
            "\nThe value of this type did not have exactly one union member specified." +
            "\nCommand name: " + commandName +
            "\nType name: " + typeName +
            "\nField name: " + fieldName +
            "\nUnion value: " + actualUnionValue
    );
  }

  public static CannotLoadRequestedDataByGraphQL becauseProvidedMutationHadParameterWhichDoesNotHaveStructureSchema(
      String commandName,
      String fieldName,
      String fieldType,
      CannotProvideStructureSchema e
  ) {
    return new CannotLoadRequestedDataByGraphQL(
        "provided mutation had parameter which does not have Structure Schema." +
            "\nCommand name: " + commandName +
            "\nParameter name: " + fieldName +
            "\nParameter type: " + fieldType,
        e
    );
  }

  public static CannotLoadRequestedDataByGraphQL becauseProvidedMutationHadParameterOfComplexTypeWithSomeFieldWhichDoesNotHaveStructureSchema(
      String commandName,
      String definitionType,
      String fieldName,
      String fieldType,
      CannotProvideStructureSchema e
  ) {
    return new CannotLoadRequestedDataByGraphQL(
        "provided mutation had parameter of complex type with some field which does not have Structure Schema."
            +
            "\nCommand name: " + commandName +
            "\nType name: " + definitionType +
            "\nField name: " + fieldName +
            "\nField type: " + fieldType,
        e
    );
  }
}
