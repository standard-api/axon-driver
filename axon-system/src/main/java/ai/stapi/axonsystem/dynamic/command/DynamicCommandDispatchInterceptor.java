package ai.stapi.axonsystem.dynamic.command;

import ai.stapi.graphsystem.messaging.command.DynamicCommand;
import java.util.List;
import java.util.function.BiFunction;
import org.axonframework.commandhandling.CommandMessage;
import org.axonframework.commandhandling.GenericCommandMessage;
import org.axonframework.messaging.GenericMessage;
import org.axonframework.messaging.MessageDispatchInterceptor;
import org.jetbrains.annotations.NotNull;

public class DynamicCommandDispatchInterceptor  implements MessageDispatchInterceptor<CommandMessage<?>> {

  @NotNull
  @Override
  public BiFunction<Integer, CommandMessage<?>, CommandMessage<?>> handle(
      @NotNull List<? extends CommandMessage<?>> messages
  ) {
    return (integer, commandMessage) -> {
      if (commandMessage.getPayload() instanceof DynamicCommand dynamicCommand) {
        return new GenericCommandMessage<>(
            new GenericMessage<>(dynamicCommand),
            dynamicCommand.getSerializationType()
        );
      }
      return commandMessage;
    };
  }
}
