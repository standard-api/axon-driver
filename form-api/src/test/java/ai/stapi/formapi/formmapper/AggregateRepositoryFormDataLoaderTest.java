package ai.stapi.formapi.formmapper;

import ai.stapi.formapi.formmapper.exceptions.CannotLoadFormData;
import ai.stapi.graphsystem.messaging.command.DynamicCommand;
import ai.stapi.graphsystem.operationdefinition.model.OperationDefinitionDTO;
import ai.stapi.graphsystem.operationdefinition.model.OperationDefinitionProvider;
import ai.stapi.identity.UniqueIdentifier;
import ai.stapi.test.domain.DomainTestCase;
import org.junit.jupiter.api.Assertions;
import org.junit.jupiter.api.Test;
import org.junit.jupiter.api.function.Executable;
import org.springframework.beans.factory.annotation.Autowired;

import java.util.List;
import java.util.Map;

class AggregateRepositoryFormDataLoaderTest extends DomainTestCase {
    
    @Autowired
    private AggregateRepositoryFormDataLoader aggregateRepositoryFormDataLoader;
    
    @Autowired
    private OperationDefinitionProvider operationDefinitionProvider;
    
    @Test
    void itThrowsExceptionWhenCommandHandlerDefinitionForProvidedOperationDoesntExist() {
        Executable throwable = () -> this.aggregateRepositoryFormDataLoader.load(
            new OperationDefinitionDTO(
                "NotExisting",
                "NotExisting",
                "Irrelevant",
                "Irrelevant",
                "Irrelevant",
                "Irrelevant",
                List.of(),
                false,
                false,
                false,
                List.of()
            ),
            "Irrelevant",
            Map.of()
        );
        this.thenExceptionMessageApproved(CannotLoadFormData.class, throwable);
    }

    @Test
    void itThrowsExceptionWhenAggregateWithProvidedResourceIdDoesntExist() {
        var operationId = "CreateParameters";
        var resourceId = "NonExistingId";
        var operation = this.operationDefinitionProvider.provide(operationId);
        Executable throwable = () -> this.aggregateRepositoryFormDataLoader.load(
            operation,
            resourceId,
            Map.of()
        );
        this.thenExceptionMessageApproved(CannotLoadFormData.class, throwable);
    }

    @Test
    void itThrowsExceptionWhenAggregateWithProvidedResourceIdDoesntxist() {
        var operationId = "CreateParameters";
        var resourceId = "SomeId";
        this.givenCommandIsDispatched(
            new DynamicCommand(
                new UniqueIdentifier(resourceId),
                operationId,
                Map.of(
                    "language", "cz",
                    "implicitRules", "http://some.uri.com"
                )
            )
        );
        var operation = this.operationDefinitionProvider.provide(operationId);
        var actual = this.aggregateRepositoryFormDataLoader.load(
            operation,
            resourceId,
            null
        );
    }
}