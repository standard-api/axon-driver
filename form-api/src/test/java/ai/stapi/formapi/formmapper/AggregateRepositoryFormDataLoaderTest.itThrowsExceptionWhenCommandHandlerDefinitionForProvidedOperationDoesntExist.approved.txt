Cannot load data for form, because there was not command handler definition with provided operation.
Operation name: 'NotExisting'